### PR TITLE
Set force on control board driver on update

### PR DESCRIPTION
**Pull request by traversaro**

### CODE REVIEW

File: include/gazebo_yarp_plugins/ControlBoardDriver.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
  * Authors: Enrico Mingo, Alessio Rocchi, Mirko Ferrati, Silvio Traversaro and Alessandro Settimi
  * CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
  */
+#include <map>
```
Comment (traversaro) [inline include/gazebo_yarp_plugins/ControlBoardDriver.h]: Spurious? I guess this is not needed anymore.

File: include/gazebo_yarp_plugins/ControlBoardDriver.h [hunk 2]
```diff
@@ -36,10 +37,13 @@ namespace yarp {
 namespace gazebo {
     namespace common {
         class UpdateInfo;
+        class PID;
     }
 
     namespace physics {
         class Model;
+        class JointController;
+        typedef boost::shared_ptr<JointController> JointControllerPtr;
```
Comment (traversaro) [inline include/gazebo_yarp_plugins/ControlBoardDriver.h]: I guess this is not needed any more either.

File: include/gazebo_yarp_plugins/ControlBoardDriver.h [hunk 3]
```diff
@@ -351,6 +355,13 @@ class yarp::dev::GazeboYarpControlBoardDriver:
     std::vector<GazeboYarpControlBoardDriver::PID> m_velocityPIDs;
     std::vector<GazeboYarpControlBoardDriver::PID> m_impedancePosPDs;
 
+
+    // std::map<std::string, gazebo::common::PID> m_gazeboPositionPIDs; /* position PID controllers for each joint, needed to call gazebo's setForce() method */
+    std::vector<gazebo::common::PID> m_gazeboPositionPIDs;
+    std::vector<gazebo::common::PID> m_gazeboVelocityPIDs; /* velocity PID controllers for each joint, needed to call gazebo's setForce() method */
+    std::vector<gazebo::common::PID> m_gazeboImpedancePosPIDs; /* impedance position PID controllers for each joint, needed to call gazebo's setForce() method */
```
Comment (traversaro) [inline include/gazebo_yarp_plugins/ControlBoardDriver.h]: I guess this is not needed, there is an internal "PID" already implemented in `void GazeboYarpControlBoardDriver::sendImpPositionToGazebo ( const int j, const double des )`

**Comment by traversaro** [comment]
@mikearagao : I thinks you should try to properly use `setForce` even for velocity and torque commands, substituting the implementation in `sendVelocityToGazebo` and `sendTorqueToGazebo` methods.

File: include/gazebo_yarp_plugins/ControlBoardDriver.h (hunk 3 above)
Comment (mikearagao) [inline include/gazebo_yarp_plugins/ControlBoardDriver.h]: @traversaro on the `void gazebo::physics::JointController::Update()` the method `gazebo::physics::Joint::SetForce(unsigned int _index, double _effort)` uses the output of `double gazebo::common::PID::Update(double _error, common::Time _dt)` as the `_effort` and I thought our solution should follow the same approach... You're saying that we should compute this `_effort` with the `GazeboYarpControlBoardDriver::PID` that already exists?

Comment (traversaro) [inline include/gazebo_yarp_plugins/ControlBoardDriver.h]: The PID is necessary for positions and velocity (`m_gazeboPositionPIDs` and `m_gazeboVelocityPIDs`) loops, that are implemented in the JointController. I was referring just to the `m_gazeboImpedancePosPIDs`, that is not needed until we replace the implementation in `sendImpPositionToGazebo`

Comment (mikearagao) [inline include/gazebo_yarp_plugins/ControlBoardDriver.h]: Ok, got it! I'll remove that one!

**Comment by mikearagao** [comment]
@traversaro do you want me to substitute those (`sendVelocityToGazebo` and `sendTorqueToGazebo`) before testing and merging the one for position?

**Comment by traversaro** [comment]
@mikearagao : I think it is better to substitute everything with our new internal loops. Otherwise I am afraid that control mode switching (as Position -> Velocity -> Torque -> Position ) would break due to incompatibility between our internal loops and the Gazebo JointController.

**Comment by mikearagao** [comment]
Ok, I'll do it asap...

**Comment by mikearagao** [comment]
Hey @traversaro! Isn't this eternal pull request a bit annoying for you there? I feel bad for letting it stay opened... Maybe we can close it and I'll go back to this as soon as possible.

A lot came in the last few weeks and I'm on a very busy schedule trying to finish my masters thesis so I have postponed this task for later.

**Comment by traversaro** [comment]
Ok @mikearagao , closing the pr for now! Feel free to update it/open a new one whenever you will work on it!

**Comment by mikearagao** [comment]
Of course! @traversaro I really am sorry about this but right now I'm in such a hurry trying to get everything ready before the deadline... Hope to be back after that.

**Comment by traversaro** [comment]
No problem, and good luck for your thesis!

**Pull request closed without merging**